mediatek: Add new spi-nand driver for mt7622/mt7629

hackpascal opened the pull request:
This PR adds a new spi-nand driver provided by MediaTek for mt7622 and mt7629 as well as all following-up SoCs.

This driver is intended to replace the existing one.
Unlike the existing snfi driver which makes use of the spi-mem framework and the spi-nand framework with modified ecc support, this driver is implemented directly on the mtd framework with other components untouched, and provides better performance, and behaves the same as the nand framework.

For mt7622 I only modified mt7622.dtsi because I don't have any of these supported devices. Since this driver does not provide the support for BMT, and BMT is not recommended for new devices, I can't modify dts files of these devices without test.

To enable this driver for mt7622, it's recommended to use new ATF & U-Boot provided by MediaTek, which are also using the new spi-nand driver:
https://github.com/mtk-openwrt/u-boot
https://github.com/mtk-openwrt/arm-trusted-firmware

OpenWrt now already supports building the new ATF & U-Boot:
Boot Loaders  --->  trusted-firmware-a-mt7622-* and u-boot-mt7622_*

To enable this driver for mt7622 kernel, please refer to mt7629-rfb.dts.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

adschm commented on target/linux/mediatek/files-5.4/drivers/mtd/mtk-snand/Kconfig:
```diff
@@ -0,0 +1,14 @@
+# Copyright (C) 2020 MediaTek Inc. All rights reserved.
+# Author: Weijie Gao <weijie.gao@mediatek.com>
+#
+# SPDX-License-Identifier: GPL-2.0
```
SPDX license identifier must be first line in the file (unless there is a shebang, then it must be second line).

Please adjust this for _all_ files in the various commits here.

hauke commented:
Is this driver on its way upstream?

981213 commented:
@hauke Definitely not. Upstream has no solution to this kind of SPI-NAND specific controllers with ECC and last time (at least a year ago) I was refered to a WIP ECC framework that still doesn't make its way into mainline kernel if I'm not mistaken.
In mediatek target, we currently misuse the spi-mem framework and hack the upstream SPI NAND driver so that it only works with this controller (in an inefficient way).
I would prefer a separated driver proposed in this PR over several hacks that breaks upstream driver for other generic spi-mem controllers.

robimarko commented:
@981213 Upstream has merged an ECC engine framework so that external ECC can be used with both RAW and SPI NAND.

981213 commented:
@robimarko This is not a spi-mem controller with an external ecc engine. Instead this controller is specifically designed to be used with spi-nand chips only. The page cache write command takes the data, calculates ECC and feed the processed data to the page cache on SPI NAND chip in one step. Also page reading gets raw data from nand chip, performs correction and return the corrected data directly.

The upstream spi-nand code performs page cache operations with spi-mem interface and ECC calculation are carried out separately with ECC framework, so that ECC engine won't work for this specific controller.

My last attempt asking upstream went nowhere and it seems I failed to explain how this controller actually works: https://www.spinics.net/lists/linux-mtd/msg11892.html
I was told in this thread that spi-mem shouldn't do anything ECC related so this controller doesn't fit into the spi-mem framework.

robimarko commented:
Ok, so I remember seeing a driver like that being pushed by Mediatek, it should be this:
https://www.spinics.net/lists/linux-mtd/msg02673.html

This was in the days when there was no SPI-NAND framework

981213 commented:
@robimarko We are using that one with several hacks. The version they sent upstream drops ECC capability and every ops performed by that driver won't mangle the data in any way. That's useless for us because we need the controllers ECC.
The hacks we did:
1. disable on-die ecc in spi-nand driver
2. enable controller ecc when it found itself executing page cache operations (They managed to left some part of this in the upstream version and that driver got NACKed immediately.)
3. read controller ecc status in spi-nand driver and report it in mtd

With all these hacks we got it up and running but this spi-nand driver won't work properly for any other spi-mem controllers.

robimarko commented:
Ok, so I think that upstream guys really don't understand why is this controller special.

dangowrt commented:
Generally I think the solution presented in this PR is the best we can have at this point (ie. as long Linux kernel upstream doesn't provide framework capable of supporting the NAND ECC, have MTD driver with built-in ECC support instead). At least it's the same driver in TF-A, U-Boot and Linux then.
The only problem I can see are devices with parallel NAND such as
[buffalo-wsr-2533dhp2](https://github.com/openwrt/openwrt/blob/master/target/linux/mediatek/dts/mt7622-buffalo-wsr-2533dhp2.dts) where we can't replace the loader easily (due to lacking support in ARM Trusted Firmware for MT7622 booting from parallel NAND) and hence got to stay with BMT/BBT stock loader...

hackpascal committed:
mediatek: add a new spi-nand driver for kernel 5.4

This patch adds a new spi-nand driver which implements the SNFI of mt7622
and mt7629.

Unlike the existing snfi driver which makes use of the spi-mem framework
and the spi-nand framework with modified ecc support, this driver is
implemented directly on the mtd framework with other components untouched,
and provides better performance, and behaves exactly the same as the nand
framework.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
mediatek: add a new spi-nand driver for kernel 5.10

This patch adds a new spi-nand driver which implements the SNFI of mt7622
and mt7629.

Unlike the existing snfi driver which makes use of the spi-mem framework
and the spi-nand framework with modified ecc support, this driver is
implemented directly on the mtd framework with other components untouched,
and provides better performance, and behaves exactly the same as the nand
framework.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
mediatek: mt7622: reserve the node name of snand for new spi-nand driver

Since we're going to add a new dts node for the new spi-nand driver with
the name 'snand', we must rename all existing nodes with 'snand' to
non-conflict ones.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
mediatek: enable new spi-nand driver for kernel 5.4

This patch enables new spi-nand driver for mt7629.
Also, the driver is unselected for mt7623 since it does not support SNFI.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal committed:
mediatek: enable new spi-nand driver for kernel 5.10

This patch enables new spi-nand driver for mt7622 and mt7629.

Only mt7622.dtsi is modified for mt7622 since this driver does not provide
the support for BMT, and BMT is not recommended for new devices.

To make use of this driver for mt7622, there are two selections:
1. Enable BMT for this driver
2. Use new ATF and U-Boot which also use this spi-nand driver:
   Boot Loaders  --->  trusted-firmware-a-mt7622-* and
   Boot Loaders  --->  u-boot-mt7622_*

To enable this driver for mt7622 kernel, please refer to mt7629-rfb.dts.

Signed-off-by: Weijie Gao <hackpascal@gmail.com>

hackpascal commented:
update: add some fixes, and move the SPDX license identifier to the first line for all files

dangowrt commented:
The only remaining problem I can see is `elecom,wrc-2533gent` which uses RealTek switch which isn't supported by U-Boot. As soon as we got that, we can replace TF-A and U-Boot there as well and get rid of BBT/BMT on SPI-NAND for good. @nbd168 ?

981213 commented:
@dangowrt 

> As soon as we got that, we can replace TF-A and U-Boot there as well

We used to NACK all device supports requiring bootloader replacement. And we probably shouldn't change that just for this BMT crap from MTK. Replacing bootloaders is not user-friendly at all and I think this kind of device support is also unmaintainable as this usually means changing partition layouts.

BMT will be part of most mt7622 devices on the market for now because no vendor bothers to remove it from MTK SDK. I think we should keep BMT support to avoid opening that can of worms. Current implementation by @nbd168 hooks into any mtd drivers with just a small patch and I think it isn't hard to port that to this driver as well.

dangowrt commented:
> @dangowrt
> 
> > As soon as we got that, we can replace TF-A and U-Boot there as well
> 
> We used to NACK all device supports requiring bootloader replacement. And we probably shouldn't change that just for this BMT crap from MTK. Replacing bootloaders is not user-friendly at all and I think this kind of device support is also unmaintainable as this usually means changing partition layouts.
> 

I don't remember we ever had such a rule. We refuse **third-party/binary/out-of-tree bootloader as a requirement** for boards added. Actually many targets require replacing the bootloader unconditionally to make any use of the hardware (think: brnboot, adam2, eva, ... on lantiq, we are always replacing that with U-Boot in order to make use of the hardware with OpenWrt). Plus lots of SBCs which only comes with romloader and U-Boot is part of our (SD Card) image (sunxi, rockchip, ...).
Replacing the bootloader once also doesn't need to be hard, it can be done using a special installer image (like https://github.com/dangowrt/linksys-e8450-openwrt-installer )

However, I agree that adding BMT/BBT by patch to this driver should not be hard and then would allow users to use OpenWrt to some degree on the stock loader and on boards where replacing the bootloader still requires more work.

pepe2k commented:
@dangowrt @981213 

> > We used to NACK all device supports requiring bootloader replacement. And we probably shouldn't change that just for this BMT crap from MTK. Replacing bootloaders is not user-friendly at all and I think this kind of device support is also unmaintainable as this usually means changing partition layouts.
> 
> I don't remember we ever had such a rule. We refuse **third-party/binary/out-of-tree bootloader as a requirement** for boards added. Actually many targets require replacing the bootloader unconditionally to make any use of the hardware (think: brnboot, adam2, eva, ... on lantiq, we are always replacing that with U-Boot in order to make use of the hardware with OpenWrt). Plus lots of SBCs which only comes with romloader and U-Boot is part of our (SD Card) image (sunxi, rockchip, ...).
> Replacing the bootloader once also doesn't need to be hard, it can be done using a special installer image (like https://github.com/dangowrt/linksys-e8450-openwrt-installer )

Side note: we usually also don't want end users to update/replace bootloaders on platforms which don't have any _easy_ way to recover if something goes wrong (e.g. UART/USB boot). But personally, the more I see what vendors do with bootloaders and locking down devices, the more I'm open to loosen this informal/unwritten requirement/restriction.

Cheers,
Piotr

dangowrt commented:
> Unfortunately they have to use UBI on top of BMT ...
What could possibly go wrong...? :-8

sysupgrade.tar would be another thing to deprecate in favor of using uImage.FIT when ever possible, but that again depends on up-to-date U-Boot. Using a single uImage.FIT for kernel, device-tree blob and rootfs has the advantage that the bootloader is then able to verify the hashes of all these components (thus avoiding reboot-loop condition in case of damaged rootfs).

Anyway, the Linux-version of the SPI-NAND driver will have to deal with legacy BMT stuff somehow. That part won't have to look nice or ever go to upstream/vanilla Linux, it can be a local-patch we keep on top of the to-be-uptreamed MTD driver and it should not be used for new boards. But we will need it to support existing systems already out there depending on it.

981213 commented:
> What could possibly go wrong...? :-8

It's unfortunate because ubi already provides bad block management and that annoying BMT is just there doing the same thing in a worse way (no write balancing).

> and it should not be used for new boards.

I think it should be used when it can significantly ease the installation procedure. In my case mentioned above, I don't want to force users to replace their bootloader (which may brick their router) when they can easily just flash a sysupgrade-tar from vendor firmware or u-boot recovery.

Anyway this discussion is a bit off-topic here. I might start a new thread when I actually submit that device support :)

nbd168 commented:
I have an idea on how to mitigate the BMT issue a bit on devices where we don't want to reflash u-boot. I could add some code to my bmt patch to exclude dts-configurable areas from remapping. We could then use a fixed size kernel partition and use the rest as UBI area.
The main complication I can think of here is that initial installation would have to be done through an initramfs image, since we can't let the vendor boot loader flash the UBI rootfs area.
We'd also still sacrifice flash space for the BMT spare area, but the extra reliability through UBI probably makes it worth doing so.
What do you think?

981213 commented:
I picked 5.10 part in this PR, added BMT support to the driver, and dropped our current spi-nand hacks from 5.10.
The result is in my staging tree: https://git.openwrt.org/?p=openwrt/staging/981213.git;a=summary
I'll probably push the changes on Friday with a bump of mt7629 subtarget to 5.10.

However it's only tested on a TOTOLINK A8000RU whose device support is still in my local branch.
It'll be great if anyone could try my staging tree on existing devices :)

dangowrt commented:
Unfortunately this breaks `factory` partition on the Linksys E8450, resulting in MT7915E not probing
```
[    8.655872] mt7915e: probe of 0000:01:00.0 failed with error -22
...
root@OpenWrt:/# hexdump -C /dev/mtdblock2
[  293.744528] blk_update_request: I/O error, dev mtdblock2, sector 0 op 0x0:(READ) flags 0x80700 phys_seg 4 prio class 0
[  293.755393] blk_update_request: I/O error, dev mtdblock2, sector 8 op 0x0:(READ) flags 0x80700 phys_seg 3 prio class 0
[  293.766235] blk_update_request: I/O error, dev mtdblock2, sector 16 op 0x0:(READ) flags 0x80700 phys_seg 2 prio class 0
[  293.777152] blk_update_request: I/O error, dev mtdblock2, sector 24 op 0x0:(READ) flags 0x80700 phys_seg 1 prio class 0
[  293.788144] blk_update_request: I/O error, dev mtdblock2, sector 0 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[  293.798571] Buffer I/O error on dev mtdblock2, logical block 0, async page read
hexdump: /dev/mtdblock2: I/O error
```

All other `/dev/mtdblock*` devices work find, even `/dev/mtd2` can be read using the character device and content looks ok.

981213 commented:
@dangowrt 
Could you try dumping mtd2 with oob data using nanddump and see what's the actual error reported?
Another idea to try: writing a factory partition backup back to mtd2 and see if that fixes the problem.
The oob layout in the hacky old snand driver is reordered, and I'm guessing maybe there are some weird data written into the bad block mark position (1st byte of oob data dumped using current driver).

dangowrt commented:
I've written back factory using `mtd` command in OpenWrt many times as I used this exact device to develop the UBI installer -- it has been wiped and rewritten (with `mtd` tool in OpenWrt or using U-Boot with the very same new SPI-NAND driver) multiple times.

Yet `nanddump` indeed shows some bit errors for `/dev/mtd2`:
```
ECC failed: 0
ECC corrected: 3264
Number of bad blocks: 0
Number of bbt blocks: 0
Block size 131072, page size 2048, OOB size 64
Dumping data starting at 0x00000000 and ending at 0x00100000...
ECC: 4 corrected bitflip(s) at offset 0x00000000
0x00000000: 22 76 .. .. .. ......
```

To compare, on /dev/mtd3 it looks like that:
```
ECC failed: 0
ECC corrected: 178
Number of bad blocks: 0
Number of bbt blocks: 0
Block size 131072, page size 2048, OOB size 64
Dumping data starting at 0x00000000 and ending at 0x07d00000...
0x00000000: 55 42 .. .. ......
```


Weirdly OOB data seems to be all `ff` everywhere *except* for `/dev/mtd2` (when trying to read it with nanddump, can be driver just doesn't allow reading it):
```
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  |................|
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  |................|
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  |................|
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  |................|
```

However, only in `/dev/mtd2` partition *all* OOB data always looks like that
```
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  |................|
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff aa aa ff  |................|
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  |................|
  OOB Data: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff  |................|
```

Obviously none of this looks like valid ECC data, but probably the _actual_ OOB data (ie. what is on the chip and not what we read through `nanddump -o`) is processed by the ECC-engine and what we see here is the _result_...

Weird enough that reading through `/dev/mtd2` works without problems and that everything was also working fine with the old spi-mem based hacky driver....

For a moment I was considering to dump the content of that SPI chip by attacking it directly via SPI -- but it's covered under a giant heat-sink which would be a huge mess to remove and I'm not even sure there is userspace tooling (ie. for use with CH341 SPI adapter) which allows dumpiung that SPI-NAND chip :frowning:

981213 commented:
@dangowrt 
I found a difference in calculation of mtd->bitflip_threshold between mtk-snand and the kernel driver:
Here's the kernel one: https://elixir.bootlin.com/linux/latest/source/drivers/mtd/nand/raw/nand_base.c#L6028
```c
	if (!mtd->bitflip_threshold)
		mtd->bitflip_threshold = DIV_ROUND_UP(mtd->ecc_strength * 3, 4);
```
and this is the one in mtk-snand: https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/mediatek/files-5.10/drivers/mtd/mtk-snand/mtk-snand-mtd.c;h=7e5baf03691126b5a30a3bdabd03bdbf325052b4;hb=HEAD#l606
```c
	mtd->ecc_strength = msm->cinfo.ecc_strength;
	mtd->bitflip_threshold = (mtd->ecc_strength * 3) / 4;
	mtd->ecc_step_size = msm->cinfo.sector_size;
```
If this is indeed a typo, it will cause mtd_read_oob to return -EUCLEAN when the number of bitflips is just at the threshold:
https://elixir.bootlin.com/linux/latest/source/drivers/mtd/mtdcore.c#L1436
```c
	if (unlikely(ret_code < 0))
		return ret_code;
	if (mtd->ecc_strength == 0)
		return 0;	/* device lacks ecc */
	return ret_code >= mtd->bitflip_threshold ? -EUCLEAN : 0;
```
we probably need someone else to test this as you've removed the bitflips by writing data back again.

981213 commented:
more investigation:
1. This problem doesn't actually affect us. ECC strength in mt7622-snand is 4 so ecc_strength * 3 / 4 = DIV_ROUND_UP(ecc_strength * 3, 4);
2. It seems that mtd_read and mtd_read_oob caller should silently ignore -EUCLEAN if they don't care about block status: https://elixir.bootlin.com/linux/v5.13.13/source/Documentation/ABI/testing/sysfs-class-mtd#L136
```
		This allows the user to examine and adjust the criteria by which
		mtd returns -EUCLEAN from mtd_read() and mtd_read_oob().  If the
		maximum number of bit errors that were corrected on any single
		region comprising an ecc step (as reported by the driver) equals
		or exceeds this value, -EUCLEAN is returned.  Otherwise, absent
		an error, 0 is returned.  Higher layers (e.g., UBI) use this
		return code as an indication that an erase block may be
		degrading and should be scrutinized as a candidate for being
		marked as bad.
```
However mt76 doesn't do that: https://elixir.bootlin.com/linux/latest/source/drivers/net/wireless/mediatek/mt76/eeprom.c#L57
and it seems that mtdblock also doesn't do that, as I'm able to reproduce this error by setting bitflip_threshold to 0:
```
root@OpenWrt:/# echo 0 > /sys/class/mtd/mtd4/bitflip_threshold 
root@OpenWrt:/# cat /dev/mtdblock4 > /dev/null
[ 9680.148708] blk_update_request: I/O error, dev mtdblock4, sector 0 op 0x0:(READ) flags 0x84700 phys_seg 32 prio class 0
[ 9680.159665] blk_update_request: I/O error, dev mtdblock4, sector 8 op 0x0:(READ) flags 0x84700 phys_seg 31 prio class 0
[ 9680.170589] blk_update_request: I/O error, dev mtdblock4, sector 16 op 0x0:(READ) flags 0x84700 phys_seg 30 prio class 0
[ 9680.181595] blk_update_request: I/O error, dev mtdblock4, sector 24 op 0x0:(READ) flags 0x84700 phys_seg 29 prio class 0
[ 9680.192622] blk_update_request: I/O error, dev mtdblock4, sector 32 op 0x0:(READ) flags 0x84700 phys_seg 28 prio class 0
[ 9680.203620] blk_update_request: I/O error, dev mtdblock4, sector 40 op 0x0:(READ) flags 0x84700 phys_seg 27 prio class 0
[ 9680.214646] blk_update_request: I/O error, dev mtdblock4, sector 48 op 0x0:(READ) flags 0x84700 phys_seg 26 prio class 0
[ 9680.225643] blk_update_request: I/O error, dev mtdblock4, sector 56 op 0x0:(READ) flags 0x84700 phys_seg 25 prio class 0
[ 9680.236641] blk_update_request: I/O error, dev mtdblock4, sector 64 op 0x0:(READ) flags 0x84700 phys_seg 24 prio class 0
[ 9680.247638] blk_update_request: I/O error, dev mtdblock4, sector 72 op 0x0:(READ) flags 0x84700 phys_seg 23 prio class 0
[ 9680.261591] Buffer I/O error on dev mtdblock4, logical block 0, async page read
[ 9680.269045] Buffer I/O error on dev mtdblock4, logical block 0, async page read
cat: read error: I/O error
root@OpenWrt:/# echo 3 > /sys/class/mtd/mtd4/bitflip_threshold 
root@OpenWrt:/# cat /dev/mtdblock4 > /dev/null
root@OpenWrt:/# 
```
I probably won't have time to prepare a fix for this during the upcoming 2 weeks though.
